tiledb/cf/engines/netcdf4_engine.py

```diff
@@ -0,0 +1,560 @@
+        from the default NetCDF file and group. These values are automatically set when
+        creating a :class:`NetCDF4ConverterEngine` using the ``from_file`` class method.
+
+        If both ``netcdf_group`` and ``input_file`` are set, this method will prioritize
+        using the NetCDF group set by ``netcdf_group``.
+
+        Parameters:
+            group_uri: Uniform resource identifier for the TileDB group data is being
+                copied to.
+            key: If not ``None``, encryption key, or dictionary of encryption keys, to
+                decrypt arrays.
+            ctx: If not ``None``, TileDB context wrapper for a TileDB storage manager.
+            input_file: If not ``None``, the NetCDF file to copy data from.
+            group_path: If not ``None``, the path to the NetCDF group to copy data from.
+            netcdf_group: If not ``None``, the NetCDF group to copy data from.
+        """
+        if netcdf_group is None:
+            input_file = (
+                input_file if input_file is not None else self._default_input_file
+            )
+            group_path = (
+                group_path if group_path is not None else self._default_group_path
+            )
+        with open_netcdf_group(netcdf_group, input_file, group_path) as nc_group:
+            # Copy group metadata
+            with Group(group_uri, mode="w", key=key, ctx=ctx) as group:
+                for group_key in nc_group.ncattrs():
+                    copy_metadata_item(group.meta, nc_group, group_key)
+            # Copy variables and variable metadata to arrays
+            for array_name, array_converter in self._arrays.items():
+                with Group(
+                    group_uri, mode="w", array=array_name, key=key, ctx=ctx
+                ) as group:
+                    data = {}
+                    if not array_converter.variables:
+                        pass
+                    for var_converter in array_converter.variables:
+                        try:
+                            variable = nc_group.variables[var_converter.input_name]
+                        except KeyError as err:
+                            raise KeyError(
+                                f"Variable {var_converter.input_name} not found in "
+                                f"requested NetCDF group."
+                            ) from err
+                        data[var_converter.output_name] = variable[...]
+                        attr_meta = group.get_attr_metadata(var_converter.output_name)
+                        for meta_key in variable.ncattrs():
+                            copy_metadata_item(attr_meta, variable, meta_key)
+                    dim_slice = tuple(slice(dim.size) for dim in variable.get_dims())
+                    if not dim_slice:
+                        group.array[:] = data
+                    else:
+                        group.array[dim_slice] = data
+
+    def get_arrays(self):
```

Comment:
I would inline (or at least privatize) this method, it doesn't have to be public.

Reply:
I don't understand what you are suggesting here. All three of these routines are used externally by DataspaceConverter class, and aren't used internally to the NetCDF4ConverterEngine. Could you clarify?

Reply:
I meant to move them in `DataspaceConverter` because (a) they're used only there and (b) they are tightly coupled with `DataspaceCreator` ("a generator of keyword arguments used to add ... to a DataspaceCreator"), not `NetCDF4ConverterEngine`.

But looking closer, moving them would imply exposing the `NetCDF4ConverterEngine.{_arrays,_dimensions}` along with their related "private" classes (`NetCDFArrayConverter`, `NetCDFDimensionConverter`), which is arguably worse.

I guess this reinforces my previous comment about using inheritance instead of composition between `DataspaceConverter` and `NetCDF4ConverterEngine`; they are too coupled to be meaningfully separated.

Reply:
I updated NetCDF4ConverterEngine to use inheritance and removed these functions altogether.